via/static/scripts/video_player/components/Transcript.tsx

```diff
@@ -45,38 +47,95 @@ export type TranscriptProps = {
    */
   currentTime: number;
 
+  /** Query string to filter the transcript. */
+  filter?: string;
+
   /**
    * Callback invoked when the user selects a segment from the transcript.
    */
   onSelectSegment?: (segment: Segment) => void;
 };
 
 type TranscriptSegmentProps = {
+  /**
+   * Mark the segment as hidden.
+   *
+   * When segments do not match the current filter, we continue to render
+   * them but hidden with CSS. This avoids losing highlights created by the
+   * Hypothesis client.
+   */
+  hidden?: boolean;
+
+  /**
+   * CSS highlight used to highlight spans of text specified by {@link matches}.
+   */
+  highlight?: Highlight;
+
+  /** Offsets within the segment text to highlight. */
+  matches?: MatchOffset[];
+
   isCurrent: boolean;
   onSelect: () => void;
   time: number;
   text: string;
 };
 
 function TranscriptSegment({
+  hidden = false,
+  highlight,
+  matches,
   isCurrent,
   onSelect,
   time,
   text,
 }: TranscriptSegmentProps) {
+  const contentRef = useRef<HTMLParagraphElement>(null);
+
+  // Highlight the text within the segment that matches the current filter
+  // query.
+  useEffect(() => {
+    if (!highlight || !matches) {
+      return () => {};
+    }
+
+    const ranges = matches.map(({ start, end }) => {
+      const textNode = contentRef.current!.childNodes[0];
+      const range = new Range();
+
+      // FIXME - If the Hypothesis client has inserted highlights, these will
+      // break up the single text node child into multiple children.
+      if (textNode instanceof Text && textNode.length >= end) {
+        range.setStart(contentRef.current!.childNodes[0], start);
+        range.setEnd(contentRef.current!.childNodes[0], end);
```

Comment:
If `textNode` is created from `contentRef.current!.childNodes[0]`, and there's already a check making sure it is a `Text` instance, could we use it in the `setStart` and `setEnd` calls?

```suggestion
        range.setStart(textNode, start);
        range.setEnd(textNode, end);
```

Reply:
Oh, indeed. I forgot to do this after extracting `contentRef.current!.childNodes[0]` into a variable.